bridge/tracker/tracker.go

```diff
@@ -0,0 +1,305 @@
+
+	// required block confirmations
+	confirmations *big.Int
+
+	// newHeads subscription channel
+	headerCh chan *ethHeader
+
+	// event channel
+	eventCh chan []byte
+
+	// cancel funcs
+	cancelSubscription, cancelEventTracking context.CancelFunc
+
+	//	rootchain client
+	client *rootchainClient
+
+	//	db for last processed block number
+	db *leveldb.DB
+}
+
+//	NewEventTracker returns a new tracker able to listen for events
+//	at the desired block depth. Events of interest are defined
+//	in rootchain.go
+func NewEventTracker(logger hclog.Logger, confirmations uint64) (*Tracker, error) {
+	tracker := &Tracker{
+		logger:        logger.Named("event_tracker"),
+		confirmations: big.NewInt(0).SetUint64(confirmations),
+		headerCh:      make(chan *ethHeader, 1),
+		eventCh:       make(chan []byte),
+	}
+
+	var err error
+
+	//	create rootchain client
+	if tracker.client, err = newRootchainClient(rootchainWS); err != nil {
+		logger.Error("cannot connect to rootchain", "err", err)
+
+		return nil, err
+	}
+
+	//	load db (last processed block number)
+	if tracker.db, err = initRootchainDB(); err != nil {
+		logger.Error("cannot initialize db", "err", err)
+
+		return nil, err
+	}
+
+	return tracker, nil
+}
+
+//	Start runs the tracker's listening mechanism:
+//	1. startEventTracking - goroutine keeping track of
+//	the last block queried, as well as collecting events
+//	of interest.
+//
+//	2. startSubscription - goroutine responsible for handling
+//	the subscription object. This object is provided by the
+//	client's subscribeNewHeads method that initiates the subscription.
+func (t *Tracker) Start() error {
+	//	create cancellable contexts
+	ctxSubscription, cancelSubscription := context.WithCancel(context.Background())
+	t.cancelSubscription = cancelSubscription
+
+	ctxEventTracking, cancelEventTracking := context.WithCancel(context.Background())
+	t.cancelEventTracking = cancelEventTracking
+
+	//	start event tracking process early
+	go t.startEventTracking(ctxEventTracking)
+
+	//	create newHeads subscription
+	subscription, err := t.client.subscribeNewHeads(t.headerCh)
+	if err != nil {
+		t.logger.Error("cannot subscribe to rootchain", "err", err)
+
+		return err
+	}
+
+	//	start receiving new header events
+	go t.startSubscription(ctxSubscription, subscription)
+
+	return nil
+}
+
+//	Stop stops the tracker's listening mechanism.
+func (t *Tracker) Stop() error {
+	//	stop subscription
+	t.cancelSubscription()
+
+	// 	stop processing headers
+	t.cancelEventTracking()
+
+	//	close rootchain client
+	if err := t.client.close(); err != nil {
+		t.logger.Error("cannot close rootchain client", "err", err)
+
+		return err
+	}
+
+	return nil
+}
+
+//	getEventChannel returns the tracker's event channel.
+func (t *Tracker) getEventChannel() <-chan []byte {
+	return t.eventCh
+}
+
+//	startSubscription handles the subscription object (provided by the rootchain client).
+func (t *Tracker) startSubscription(ctx context.Context, sub subscription) {
```

Comment:
`startSubscription` is not proper name...? because this function just handles error or terminate signals, not process any message. And we don't need for loop here, right?

Reply:
[Refactored](https://github.com/0xPolygon/polygon-edge/pull/386/commits/22ed210d9e92dd02820edc89fd3ec5dc0e6cad92)